File: app/src/main/java/ua/com/amicablesoft/android/wr/ui/MainPresenter.java
package ua.com.amicablesoft.android.wr.ui;

import android.os.Environment;

import com.google.firebase.auth.FirebaseAuth;

import java.io.File;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

import ua.com.amicablesoft.android.wr.R;
import ua.com.amicablesoft.android.wr.dal.IRepository;
import ua.com.amicablesoft.android.wr.dal.Repository;
import ua.com.amicablesoft.android.wr.models.Exercise;
import ua.com.amicablesoft.android.wr.models.Powerlifter;
import ua.com.amicablesoft.android.wr.models.User;

/**
 * Created by lapa on 25.09.16.
 */

class MainPresenter {

    private final MainView view;
    private Powerlifter currentPowerlifter;
    private Exercise currentExercise;
    private Repository repository;
    private File currentVideoPath;

    MainPresenter (MainView view) {
        this.view = view;
    }

    void onStart() {
        view.showLoading();
        repository = new Repository();
        repository.getPowerlifters(new IRepository.LoadPowerliftersCallback() {
            @Override
            public void onPowerliftersLoaded(ArrayList<Powerlifter> powerlifters) {
                view.setListPowerlifters(powerlifters);
                if (powerlifters.isEmpty()) {
                    view.openAddPowerlifterView();
                }
                view.dismissLoading();
            }

            @Override
            public void onDataNotAvailable() {
                view.dismissLoading();
            }
        });
        view.setPowerlifter(0);
        view.setExercise(1);
        currentExercise = Exercise.BenchPress;
    }

    void onPowerlifterAdded() {
        repository.getPowerlifters(new IRepository.LoadPowerliftersCallback() {
            @Override
            public void onPowerliftersLoaded(ArrayList<Powerlifter> powerlifters) {
                view.setListPowerlifters(powerlifters);
            }

            @Override
            public void onDataNotAvailable() {

            }
        });
        view.setPowerlifter(0);
    }

    void changePowerlifter(Powerlifter powerlifter) {
        currentPowerlifter = powerlifter;
    }

    void changeExercise(Integer exercise) {
        if (exercise == 0) {
            currentExercise = Exercise.Squats;
        } else if (exercise == 1) {
            currentExercise = Exercise.BenchPress;
        } else if (exercise == 2) {
            currentExercise = Exercise.DeadLift;
        }
    }

    private String createVideoFileName() throws IOException {
        String date = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(new Date());
        String name = getPowerlifterName();
        String exercise = currentExercise.toString();
        String setNumber = getSetNumber();
        return name + "-" + exercise + "-" + date + "-" + setNumber + ".mp4";
    }

    private String createDirName() {
        if (currentPowerlifter != null) {
            String powerlifterName = currentPowerlifter.getLastName() + "-" + currentPowerlifter.getName();
            String exercise = currentExercise.toString();
            String date = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(new Date());
            return "/" + powerlifterName + "/" + exercise + "/" + date + "/";
        } else {
            view.showSnackbar(R.string.snackbar_text_powerlifter);
            return null;
        }
    }

    private File createFile() throws IOException {
        String dirName = createDirName();
        File filePath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES);
        File videoPath = new File(filePath, dirName);
        videoPath.mkdirs();
        currentVideoPath = videoPath;
        String fileName = createVideoFileName();
        return new File(videoPath, fileName);
    }

    void onAuthInSuccess() {
        final Repository repository = new Repository();
        FirebaseAuth auth = FirebaseAuth.getInstance();
        final User user = new User(auth.getCurrentUser().getUid(), auth.getCurrentUser().getEmail());
        repository.userExist(user, new IRepository.LoadUserCallback() {
                    @Override
                    public void found(boolean userFound) {
                        if (!userFound) {
                            repository.userSave(user);
                        }
                    }
                });
    }

    void onNewVideoAction() {
        view.requestPermissions();
    }

    void onSnackbarVideoAction() {
        view.showSnackbar(R.string.snackbar_text_video);
    }

    void onSnackbarPermissions() {
        view.showSnackbar(R.string.snackbar_text_permissions);
    }

    void createVideo() throws IOException {
        File videoFile = createFile();
        view.recordVideo(videoFile);
    }

    private String getPowerlifterName() {
        String name = currentPowerlifter.getName();
        String lastName = currentPowerlifter.getLastName();
        return lastName.substring(0, 1) + name.substring(0, 1);
    }

    private int getNumberOfFiles() {
        int count = 0;
        if (currentVideoPath != null) {
            count = currentVideoPath.listFiles().length;
        }
        return count;
    }

    private String getSetNumber() {
        int count = getNumberOfFiles();
        Integer number = count + 1;
        return number.toString();
    }
}

File: app/src/main/java/ua/com/amicablesoft/android/wr/ui/main/MainPresenter.kt
/*
 * Created by Olha Tymoshenko on 5/25/18 1:43 PM.
 * Copyright (c) 2018. All rights reserved.
 */

package ua.com.amicablesoft.android.wr.ui.main

import android.os.Environment
import com.google.firebase.auth.FirebaseAuth
import ua.com.amicablesoft.android.wr.R
import ua.com.amicablesoft.android.wr.dal.IRepository
import ua.com.amicablesoft.android.wr.dal.Repository
import ua.com.amicablesoft.android.wr.models.Competition
import ua.com.amicablesoft.android.wr.models.Exercise
import ua.com.amicablesoft.android.wr.models.Powerlifter
import ua.com.amicablesoft.android.wr.models.User
import java.io.File
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*

internal class MainPresenter(private val view: MainView) {
    private var currentPowerlifter: Powerlifter? = null
    private var currentCompetition: Competition? = null
    private var currentExercise: Exercise? = null
    private var repository: Repository? = null
    private var currentVideoPath: File? = null

    private val powerlifterName: String
        get() {
            val name = currentPowerlifter!!.name
            val lastName = currentPowerlifter!!.lastName
            return lastName!!.substring(0, 1) + name!!.substring(0, 1)
        }

    private val numberOfFiles: Int
        get() {
            var count = 0
            if (currentVideoPath != null) {
                count = currentVideoPath!!.listFiles().size
            }
            return count
        }

    private val setNumber: String
        get() {
            val count = numberOfFiles
            val number = count + 1
            return number.toString()
        }

    fun onStart() {
        view.showLoading()
        onAuthInSuccess()
        repository = Repository()
        repository!!.getPowerlifters(object : IRepository.LoadPowerliftersCallback {
            override fun onPowerliftersLoaded(powerlifterArrayList: List<Powerlifter?>) {
                view.setListPowerlifters(powerlifterArrayList as ArrayList<Powerlifter>)
                if (powerlifterArrayList.isEmpty()) {
                    view.openAddPowerlifterView()
                }
                view.dismissLoading()
            }

            override fun onDataNotAvailable() {
                view.dismissLoading()
            }
        })
        getCompetitionsFromRepository()
        view.setPowerlifter(0)
        view.setExercise(1)
        currentExercise = Exercise.BENCHPRESS
    }

    fun onPowerlifterAdded() {
        repository!!.getPowerlifters(object : IRepository.LoadPowerliftersCallback {
            override fun onPowerliftersLoaded(powerlifterArrayList: List<Powerlifter?>) {
                view.setListPowerlifters(powerlifterArrayList as ArrayList<Powerlifter>)
            }

            override fun onDataNotAvailable() {

            }
        })
        view.setPowerlifter(0)
    }

    fun onCompetitionAdded() {
        getCompetitionsFromRepository()
    }

    private fun getCompetitionsFromRepository() {
        repository!!.getCompetitions(object : IRepository.LoadCompetitionsCallback {
            override fun onCompetitionsLoaded(competitionArrayList: List<Competition?>) {
                view.setListCompetitions(competitionArrayList as ArrayList<Competition>)
            }

            override fun onDataNotAvailable() {

            }
        })
        view.setCompetition(0)
    }

    fun changePowerlifter(powerlifter: Powerlifter) {
        currentPowerlifter = powerlifter
    }

    fun changeCompetition(competition: Competition) {
        currentCompetition = competition
    }

    fun changeExercise(exercise: Int?) {
        if (exercise == 0) {
            currentExercise = Exercise.SQUATS
        } else if (exercise == 1) {
            currentExercise = Exercise.BENCHPRESS
        } else if (exercise == 2) {
            currentExercise = Exercise.DEADLIFT
        }
    }

    @Throws(IOException::class)
    private fun createVideoFileName(): String {
        val date = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())
        val name = powerlifterName
        val exercise = currentExercise!!.toString()
        val setNumber = setNumber
        return "$name-$exercise-$date-$setNumber.mp4"
    }

    private fun createDirName(): String? {
        if (currentPowerlifter != null) {
            val competition = currentCompetition!!.competition
            val powerlifterName = currentPowerlifter!!.lastName + "-" + currentPowerlifter!!.name
            val exercise = currentExercise!!.toString()
            val date = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault()).format(Date())
            return if (!competition!!.contentEquals("- None -")) {
                "/$competition/$powerlifterName/$exercise/$date/"
            } else {
                "/$powerlifterName/$exercise/$date/"
            }
        } else {
            view.showSnackbar(R.string.snackbar_text_powerlifter)
            return null
        }
    }

    @Throws(IOException::class)
    private fun createFile(): File {
        val dirName = createDirName()
        val filePath = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_MOVIES)
        val videoPath = File(filePath, dirName!!)
        videoPath.mkdirs()
        currentVideoPath = videoPath
        val fileName = createVideoFileName()
        return File(videoPath, fileName)
    }

    private fun onAuthInSuccess() {
        val repository = Repository()
        val auth = FirebaseAuth.getInstance()
        val user = User(auth.currentUser!!.uid, auth.currentUser!!.email!!)
        repository.userExist(user, object : IRepository.LoadUserCallback {
            override fun found(userFound: Boolean?) {
                if ((!userFound!!)) {
                    repository.userSave(user)
                }
            }
        })
    }

    fun onNewVideoAction() {
        view.requestPermissions()
    }

    fun onSnackbarVideoAction() {
        view.showSnackbar(R.string.snackbar_text_video)
    }

    fun onSnackbarPermissions() {
        view.showSnackbar(R.string.snackbar_text_permissions)
    }

    @Throws(IOException::class)
    fun createVideo() {
        val videoFile = createFile()
        view.recordVideo(videoFile)
    }
}
